File: src/main/java/org/cloudio/morpheus/mail/DefaultUserMail.java
package org.cloudio.morpheus.mail;

/**
 * Created by zslajchrt on 24/08/15.
 */
public class DefaultUserMail implements UserMail {

    private final MailOwner mailOwner;

    public DefaultUserMail(MailOwner mailOwner) {
        this.mailOwner = mailOwner;
    }

    public void sendEmail(Message message) {
        try {
            validateEmail(message);
            send(message);
        } catch (IllegalArgumentException e) {
            store(message);
        }
    }

    public void validateEmail(Message message) {
        // todo
    }

    private void send(Message message) {
        // todo
    }

    private void store(Message message) {
        // todo
    }

    public MailOwner getMailOwner() {
        return mailOwner;
    }
}

File: src/main/scala/org/cloudio/morpheus/mail/traditional/DefaultUserMail.scala
package org.cloudio.morpheus.mail.traditional

import org.cloudio.morpheus.mail.MailOwner

/**
 * Created by zslajchrt on 24/08/15.
 */
trait DefaultUserMail extends UserMail {

  this: MailOwner =>

  override def sendEmail(message: Message): Unit = {
    try {
      validateEmail(message)
      send(message)
    }
    catch {
      case e: IllegalArgumentException =>
        store(message)
    }
  }

  override def validateEmail(message: Message): Unit = {
    // todo
  }

  private def send(message: Message): Unit = {
    // todo
  }

  private def store(message: Message): Unit = {
    // todo
  }

}
